test/test_compile.py

```diff
@@ -0,0 +1,66 @@
+# Copyright Contributors to the Pyro project.
+# SPDX-License-Identifier: Apache-2.0
+
+import pytest
+
+from jax import lax, random
+from jax.lib import xla_bridge
+
+import numpyro
+from numpyro.contrib.autoguide import AutoDiagonalNormal
+import numpyro.distributions as dist
+from numpyro.infer import ELBO, MCMC, NUTS, SVI
+import numpyro.optim as optim
+
+
+GLOBAL = {"count": 0}
+
+
+def model(deterministic=True):
+    GLOBAL["count"] += 1
+    x = numpyro.sample("x", dist.Normal())
+    if deterministic:
+        numpyro.deterministic("x_copy", x)
+
+
+@pytest.mark.parametrize('deterministic', [True, False])
+def test_mcmc_one_chain(deterministic):
+    GLOBAL["count"] = 0
+    mcmc = MCMC(NUTS(model), 100, 100)
+    mcmc.run(random.PRNGKey(0), deterministic=deterministic)
+    mcmc.get_samples()
+
+    if deterministic:
+        assert GLOBAL["count"] == 8
+    else:
+        assert GLOBAL["count"] == 7
```

Comment:
@eb8680 @neerajprad this is pretty bad; with recent refactoring, in principle, we should just need to travel through the model 3 times:
+ 1 to inspect the model to get information such as prototype params, inverse transforms, has deterministic sites or not, using enumerate or not,...
+ 1 to find initial params
+ 1 to run mcmc loops

Reply:
We are traversing through the model 7 times, but are we also calling into `jit` 7 times? The latter is what we are interested in optimizing.

Reply:
FYI..not for testing, but for local debugging you can use the jax flag `jax.flags.FLAGS['log_jax_compiles'] = True` to log every time the model is being compiled.

Reply:
For single chain, the body_fn of find_valid_initial_params travels through the model 2 times. The sample_fn of hmc travels through the model 2 times, so the situation might be not that bad. But for multi-chains, I think we run pmap over about 6 or 7 traces, which is pretty bad.

Reply:
Thanks for explaining, that makes sense.